plutus-scb/src/Plutus/SCB/Core/ContractInstance.hs

```diff
@@ -0,0 +1,484 @@
+    , Member ChainIndexEffect effs
+    , Member Log effs
+    )
+    => Eff effs ()
+processTxConfirmedRequests = do
+    logInfo "processTxConfirmedRequests start"
+    state <- Query.contractStates <$> runGlobalQuery (Query.awaitTxConfirmedRequests @t)
+    traverse_ (uncurry $ processTxConfirmedRequest @t) (Map.toList state)
+    logInfo "processTxConfirmedRequests end"
+
+processTxConfirmedRequest ::
+    forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member ChainIndexEffect effs
+    , Member Log effs
+    )
+    => ContractInstanceId
+    -> (ContractIteration, TxIdSet)
+    -> Eff effs [ChainEvent t]
+processTxConfirmedRequest i (it, TxIdSet txids) =
+    let go txid = do
+            logDebug . render $ "processTxConfirmedRequest" <+> pretty i <+> pretty txid
+            confirmed <- transactionConfirmed txid
+            if confirmed
+                then sendContractInboxMessage (AwaitTxConfirmedResponse $ TxConfirmed txid) i it
+                else pure []
+    in fold <$> traverse go (Set.toList txids)
+
+-- | Look at the outboxes of all contract instances and process them.
+processAllContractOutboxes ::
+    forall t effs.
+    ( Member Log effs
+    , Member (EventLogEffect (ChainEvent t)) effs
+    , Member WalletEffect effs
+    , Member ChainIndexEffect effs
+    , Member SigningProcessEffect effs
+    )
+    => Eff effs ()
+processAllContractOutboxes = do
+    logInfo "processAllContractOutboxes start"
+    -- TODO: maybe this should be interleaved with 'processAllContractInboxes'
+    --       to avoid doing too much work?
+    processOwnPubkeyRequests @t
+    processAwaitSlotRequests @t
+    processUtxoAtRequests @t
+    processWriteTxRequests @t
+    processTxConfirmedRequests @t
+    processNextTxAtRequests @t
+    logInfo "processAllContractOutboxes end"
+
+callContractEndpoint ::
+    forall t a effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member Log effs
+    , Member (Error SCBError) effs
+    , JSON.ToJSON a
+    )
+    => ContractInstanceId
+    -> String
+    -> a
+    -> Eff effs [ChainEvent t]
+callContractEndpoint inst endpointName endpointValue = do
+    logInfo . render $ "calling endpoint" <+> pretty endpointName <+> "on instance" <+> pretty inst
+    state <- Query.contractStates <$> runGlobalQuery (Query.userEndpointRequests @t)
+    (it, ActiveEndpoints activeEndpointNames) <-
+        maybe
+            (throwError (OtherError $ "Contract " <> Text.pack (show inst) <> " has not requested any user input"))
+            pure
+            (Map.lookup inst state)
+    if (EndpointDescription endpointName) `Set.member` activeEndpointNames
+        then
+            let rsp = UserEndpointResponse (Text.pack endpointName) (EndpointValue $ JSON.toJSON endpointValue)
+            in sendContractInboxMessage @t rsp inst it
+        else throwError (OtherError $ "Contract " <> Text.pack (show inst) <> " has not requested endpoint " <> Text.pack endpointName)
```

Comment:
For stuff like this I'd be tempted to use the shortcircuiting behaviour of errors to do
```
unless condition $ throwError ...
let rsp = ...
```
Saves a little indentation.

Reply:
Good point.